### PR TITLE
Cash 513

### PR DESCRIPTION
**Pull request by BoulderBrains**

### CODE REVIEW

File: src/components/LoanCards/FavoriteStar.vue [hunk 1]
```diff
@@ -26,10 +29,58 @@ export default {
 	},
 	methods: {
 		toggleFavorite() {
-			this.$emit('favorite-toggled');
+			if (this.isFavorite === false) {
```
Comment (mcstover) [inline src/components/LoanCards/FavoriteStar.vue]: Spoke with @BoulderBrains about incorporating this into GridLoanCard and LoanCardController instead.

File: src/components/LoanCards/GridLoanCard.vue [hunk 2]
```diff
@@ -196,24 +197,36 @@ export default {
 	methods: {
 		toggleFavorite() {
 			// optimistically toggle it locally first
-			this.isFavorite = !this.isFavorite;
+			// this.isFavorite = !this.isFavorite;
```
Comment (mcstover) [inline src/components/LoanCards/GridLoanCard.vue]: Prob should keep this. It sets the "optimistic value" and immediately shows the new expected star state.

File: src/components/LoanCards/GridLoanCard.vue [hunk 3]
```diff
@@ -196,24 +197,36 @@ export default {
...
-					if (favorite === null) {
-						this.isFavorite = !this.isFavorite;
+			}).then(data => {
+				if (data.errors) {
+					_forEach(data.errors, ({ message }) => {
+						this.$showTipMsg(message, 'error');
+					});
+				} else {
+					this.$kvTrackEvent(
+						'favorited',
```
Comment (mcstover) [inline src/components/LoanCards/GridLoanCard.vue]: This should be "Lending" to put it into the context of other lending flow events.

File: src/components/LoanCards/GridLoanCard.vue [hunk 4]
```diff
@@ -196,24 +197,36 @@ export default {
...
-					const favorite = _get(data, 'loan.favorite');
-
-					if (favorite === null) {
-						this.isFavorite = !this.isFavorite;
+			}).then(data => {
+				if (data.errors) {
+					_forEach(data.errors, ({ message }) => {
+						this.$showTipMsg(message, 'error');
+					});
+				} else {
```
Comment (mcstover) [inline src/components/LoanCards/GridLoanCard.vue]: You probably want to also set this.isFavorite just inside this else statement using the return value from your mutation "data".

File: src/components/LoanCards/GridLoanCard.vue [hunk 5]
```diff
@@ -196,24 +197,36 @@ export default {
...
+					);
+					if (this.isFavorite === true) {
+						// eslint-disable-next-line max-len
+						this.$showTipMsg('This loan has been saved to your "Starred loans" list, which is accessible under the "Lend" menu in the header.', 'confirm');
 					}
 				}
+			}).catch(error => {
+				console.error(error);
 			});
+			this.isFavorite = !this.isFavorite;
```
Comment (mcstover) [inline src/components/LoanCards/GridLoanCard.vue]: this won't be necessary with the changes mentioned above.

**Comment by BoulderBrains** [comment]
Alright, @mcstover I've made some changes after our discussion today and this is now working again. A couple things I still want to talk through with you. 

1) The comments I left on lines 172&173 of GridLoanCard.vue. That toggle really feels like it should happen only after we're not getting an error response from graphql. 

2) I've noticed in some situations we have the same loan card rendered on the page twice. WHen I favorite one of those loans, the GridLoanCard I clicked on to favorite updates, but the other copy of the same card stays unchecked until the page refreshes.

**Comment by BoulderBrains** [comment]
This new branch has been created for the improvement of the favorite functionality work that we discussed through the course of this work: https://github.com/kiva/ui/tree/work-toward-improved-favoriting

**Pull request merged**